types/block.go

```diff
@@ -434,68 +434,129 @@ func (h *Header) StringIndented(indent string) string {
 
 //-------------------------------------
 
-// CommitSig is a vote included in a Commit.
-// For now, it is identical to a vote,
-// but in the future it will contain fewer fields
-// to eliminate the redundancy in commits.
-// See https://github.com/tendermint/tendermint/issues/1648.
-type CommitSig Vote
+// BlockIDFlag indicates which BlockID the signature is for.
+type BlockIDFlag byte
 
-// String returns the underlying Vote.String()
-func (cs *CommitSig) String() string {
-	return cs.toVote().String()
+const (
+	// BlockIDFlagAbsent - vote is not included in the Commit.Precommits.
+	BlockIDFlagAbsent BlockIDFlag = iota + 1
+	// BlockIDFlagCommit - voted for the Commit.BlockID.
+	BlockIDFlagCommit
+	// BlockIDFlagNil - voted for nil.
+	BlockIDFlagNil
+)
+
+// CommitSig is a part of the Vote included in a Commit.
+type CommitSig struct {
+	BlockIDFlag      BlockIDFlag `json:"block_id_flag"`
+	ValidatorAddress Address     `json:"validator_address"`
+	Timestamp        time.Time   `json:"timestamp"`
+	Signature        []byte      `json:"signature"`
 }
 
-// toVote converts the CommitSig to a vote.
-// TODO: deprecate for #1648. Converting to Vote will require
-// access to ValidatorSet.
-func (cs *CommitSig) toVote() *Vote {
-	if cs == nil {
-		return nil
+// NewCommitSigForBlock returns new CommitSig with BlockIDFlagCommit.
+func NewCommitSigForBlock(signature []byte, valAddr Address, ts time.Time) CommitSig {
+	return CommitSig{
+		BlockIDFlag:      BlockIDFlagCommit,
+		ValidatorAddress: valAddr,
+		Timestamp:        ts,
+		Signature:        signature,
 	}
-	v := Vote(*cs)
-	return &v
+}
+
+func (cs CommitSig) String() string {
+	return fmt.Sprintf("CommitSig{%X by %X on %v @ %s}",
+		cmn.Fingerprint(cs.Signature),
+		cmn.Fingerprint(cs.ValidatorAddress),
+		cs.BlockIDFlag,
+		CanonicalTime(cs.Timestamp))
+}
+
+// BlockID returns the Commit's BlockID if CommitSig indicates signing,
+// otherwise - empty BlockID.
+func (cs CommitSig) BlockID(commitBlockID BlockID) BlockID {
+	var blockID BlockID
+	switch cs.BlockIDFlag {
+	case BlockIDFlagAbsent:
+		blockID = BlockID{}
+	case BlockIDFlagCommit:
+		blockID = commitBlockID
+	case BlockIDFlagNil:
+		blockID = BlockID{}
+	default:
+		panic(fmt.Sprintf("Unknown BlockIDFlag: %v", cs.BlockIDFlag))
+	}
+	return blockID
+}
+
+// ValidateBasic performs basic validation.
+func (cs CommitSig) ValidateBasic() error {
+	switch cs.BlockIDFlag {
+	case BlockIDFlagAbsent:
+	case BlockIDFlagCommit:
+	case BlockIDFlagNil:
+	default:
+		return fmt.Errorf("unknown BlockIDFlag: %v", cs.BlockIDFlag)
+	}
+
+	// if len(cs.ValidatorAddress) != crypto.AddressSize {
+	// 	return fmt.Errorf("expected ValidatorAddress size to be %d bytes, got %d bytes",
+	// 		crypto.AddressSize,
+	// 		len(cs.ValidatorAddress),
+	// 	)
+	// }
+
+	// NOTE: Timestamp validation is subtle and handled elsewhere.
+
+	// if len(cs.Signature) == 0 {
+	// 	return errors.New("signature is missing")
+	// }
```

Comment:
do you want to remove this commented-out section?

Reply:
will do